libraries/AP_Compass/Compass.cpp

```diff
@@ -424,45 +424,41 @@ void Compass::_detect_backends(void)
 #elif HAL_COMPASS_DEFAULT == HAL_COMPASS_BH
     // detect_mpu9250() failed will cause panic if no actual mpu9250 backend,
     // in BH, only one compass should be detected
-    AP_Compass_Backend *backend = AP_Compass_HMC5843::detect_i2c(*this, hal.i2c);
+    AP_Compass_Backend *backend = AP_Compass_HMC5843::probe(*this, hal.i2c_mgr->get_device(HAL_COMPASS_HMC5843_I2C_BUS, HAL_COMPASS_HMC5843_I2C_ADDR));
     if (backend) {
         _add_backend(backend);
     } else {
         _add_backend(AP_Compass_AK8963::detect_mpu9250(*this, 0));
     }
-#elif CONFIG_HAL_BOARD == HAL_BOARD_LINUX && \
-      CONFIG_HAL_BOARD_SUBTYPE != HAL_BOARD_SUBTYPE_LINUX_NONE && \
-      CONFIG_HAL_BOARD_SUBTYPE != HAL_BOARD_SUBTYPE_LINUX_BEBOP && \
-      CONFIG_HAL_BOARD_SUBTYPE != HAL_BOARD_SUBTYPE_LINUX_QFLIGHT && \
-      CONFIG_HAL_BOARD_SUBTYPE != HAL_BOARD_SUBTYPE_LINUX_MINLURE
-    _add_backend(AP_Compass_HMC5843::detect_i2c(*this, hal.i2c));
+#elif HAL_COMPASS_DEFAULT == HAL_COMPASS_QFLIGHT
+    _add_backend(AP_Compass_QFLIGHT::detect(*this));
+#elif HAL_COMPASS_DEFAULT == HAL_COMPASS_AK8963_MPU9250
     _add_backend(AP_Compass_AK8963::detect_mpu9250(*this, 0));
-#elif HAL_COMPASS_DEFAULT == HAL_COMPASS_HIL
-    _add_backend(AP_Compass_HIL::detect(*this));
 #elif HAL_COMPASS_DEFAULT == HAL_COMPASS_HMC5843
-    _add_backend(AP_Compass_HMC5843::detect_i2c(*this, hal.i2c));
+    _add_backend(AP_Compass_HMC5843::probe(*this, hal.i2c_mgr->get_device(HAL_COMPASS_HMC5843_I2C_BUS, HAL_COMPASS_HMC5843_I2C_ADDR)));
 #elif HAL_COMPASS_DEFAULT == HAL_COMPASS_HMC5843_MPU6000
-    _add_backend(AP_Compass_HMC5843::detect_mpu6000(*this));
-#elif  HAL_COMPASS_DEFAULT == HAL_COMPASS_AK8963_I2C && HAL_INS_AK8963_I2C_BUS == 1
+    _add_backend(AP_Compass_HMC5843::probe_mpu6000(*this));
+#elif  HAL_COMPASS_DEFAULT == HAL_COMPASS_AK8963_I2C
     _add_backend(AP_Compass_AK8963::detect_i2c(*this, hal.i2c1,
                                                HAL_COMPASS_AK8963_I2C_ADDR));
 #elif HAL_COMPASS_DEFAULT == HAL_COMPASS_AK8963_MPU9250_I2C
     _add_backend(AP_Compass_AK8963::detect_mpu9250_i2c(*this, HAL_COMPASS_AK8963_I2C_POINTER,
                                                        HAL_COMPASS_AK8963_I2C_ADDR));
-#elif HAL_COMPASS_DEFAULT == HAL_COMPASS_PX4 || HAL_COMPASS_DEFAULT == HAL_COMPASS_VRBRAIN
-    _add_backend(AP_Compass_PX4::detect(*this));
-#elif HAL_COMPASS_DEFAULT == HAL_COMPASS_AK8963_MPU9250
+#elif CONFIG_HAL_BOARD == HAL_BOARD_LINUX && CONFIG_HAL_BOARD_SUBTYPE != HAL_BOARD_SUBTYPE_LINUX_NONE
+    _add_backend(AP_Compass_HMC5843::probe(*this, hal.i2c_mgr->get_device(HAL_COMPASS_HMC5843_I2C_BUS, HAL_COMPASS_HMC5843_I2C_ADDR)));
     _add_backend(AP_Compass_AK8963::detect_mpu9250(*this, 0));
-#elif HAL_COMPASS_DEFAULT == HAL_COMPASS_QFLIGHT
-    _add_backend(AP_Compass_QFLIGHT::detect(*this));
-#elif HAL_COMPASS_DEFAULT == HAL_COMPASS_QURT
-    _add_backend(AP_Compass_QURT::detect(*this));
 #else
     #error Unrecognised HAL_COMPASS_TYPE setting
 #endif
```

Comment:
> I think it would be easier to review if this commit was split.

github pretends to fix its bad UI for review by making a worse one. Now I don't know what commit you are talking about.

> It would be nice to have an explanation on why not using _retry_time on general register writes and reads.

Retry in general is a bad design.  It should be way up in the decision making, not in the lower level code.  In this particular case I moved the retry to be done only while collecting samples.  If you can't pass the initialization phase without resorting to "retry", go home and fix your cabling/sensors - and... reducing the bus speed affects **all** users in that bus and can be an even worse problem than just setting the sensor as unhealthy.